File: android/src/main/java/com/cappitolian/plugins/httplocalserviceswifter/HttpLocalServerSwifterPlugin.java
package com.cappitolian.plugins.httplocalserviceswifter;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "HttpLocalServerSwifter")
public class HttpLocalServerSwifterPlugin extends Plugin {
    private HttpLocalServerSwifter localServer;

    @Override
    public void load() {
        super.load();
        localServer = new HttpLocalServerSwifter(this);
    }

    @PluginMethod
    public void connect(PluginCall call) {
        localServer.connect(call);
    }

    @PluginMethod
    public void disconnect(PluginCall call) {
        localServer.disconnect(call);
    }

    @PluginMethod
    public void sendResponse(PluginCall call) {
        String requestId = call.getString("requestId");
        if (requestId == null || requestId.isEmpty()) {
            call.reject("Missing requestId");
            return;
        }
        
        // Pass the entire PluginCall data object to handleJsResponse
        // This allows us to capture 'status' and 'headers' along with the 'body'
        HttpLocalServerSwifter.handleJsResponse(requestId, call.getData());
        call.resolve();
    }

    public void fireOnRequest(JSObject data) {
        notifyListeners("onRequest", data);
    }
}

File: ios/Sources/HttpLocalServerSwifterPlugin/HttpLocalServerSwifterPlugin.swift
import Foundation
import Capacitor

@objc(HttpLocalServerSwifterPlugin)
public class HttpLocalServerSwifterPlugin: CAPPlugin, CAPBridgedPlugin, HttpLocalServerSwifterDelegate {
    public let identifier   = "HttpLocalServerSwifterPlugin"
    public let jsName       = "HttpLocalServerSwifter"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "connect",      returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "disconnect",   returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sendResponse", returnType: CAPPluginReturnPromise)
    ]

    private var localServer: HttpLocalServerSwifter?

    // MARK: - Plugin methods

    @objc func connect(_ call: CAPPluginCall) {
        if localServer == nil {
            localServer = HttpLocalServerSwifter(delegate: self)
        }
        localServer?.connect(call)
    }

    @objc func disconnect(_ call: CAPPluginCall) {
        // Use the dedicated method that accepts a call to resolve so the
        // ambiguous no-argument overload is gone entirely.
        localServer?.disconnect(resolving: call)
        localServer = nil
    }

    @objc func sendResponse(_ call: CAPPluginCall) {
        guard let requestId = call.getString("requestId"), !requestId.isEmpty else {
            call.reject("Missing requestId")
            return
        }

        let responseData = call.dictionaryRepresentation as? [String: Any] ?? [:]
        
        HttpLocalServerSwifter.handleJsResponse(
            requestId:    requestId,
            responseData: responseData
        )
        call.resolve()
    }

    // MARK: - Delegate

    public func httpLocalServerSwifterDidReceiveRequest(_ data: [String: Any]) {
        notifyListeners("onRequest", data: data)
    }
}